File: src/domain/Pattern.cpp
#include "flexseq/Pattern.h"

namespace flexseq {

bool isValidRatchet(uint8_t code) {
    return code == RATCHET_NONE || code == RATCHET_2 || code == RATCHET_3 ||
           code == RATCHET_4 || code == RATCHET_6 || code == RATCHET_TRIPLET;
}

uint8_t ratchetTriggers(uint8_t code) {
    if (code == RATCHET_TRIPLET) {
        return 3; // three triggers, spread over two step durations
    }
    if (code == RATCHET_2 || code == RATCHET_3 || code == RATCHET_4 ||
        code == RATCHET_6) {
        return code; // the code IS the trigger count
    }
    return 1;
}

uint8_t ratchetSpan(uint8_t code) {
    return (code == RATCHET_TRIPLET) ? 2 : 1;
}

bool ratchetFitsStep(uint8_t code, uint16_t ticksPerStep) {
    const uint8_t triggers = ratchetTriggers(code);
    if (triggers <= 1) {
        return true;
    }
    const uint32_t stepTicks =
        static_cast<uint32_t>(ticksPerStep) * ratchetSpan(code);
    return stepTicks / triggers >= MIN_SLOT_TICKS;
}

bool Pattern::readStep(uint8_t index, bool& active) const {
    if (index >= DEFAULT_TOTAL_STEPS) {
        return false;
    }

    const uint8_t byteIndex = index >> 3;
    const uint8_t bitIndex = index & 0x07;
    const uint8_t mask = static_cast<uint8_t>(1u << bitIndex);

    active = (packedSteps[byteIndex] & mask) != 0;
    return true;
}

bool Pattern::writeStep(uint8_t index, bool active) {
    if (index >= DEFAULT_TOTAL_STEPS) {
        return false;
    }

    const uint8_t byteIndex = index >> 3;
    const uint8_t bitIndex = index & 0x07;
    const uint8_t mask = static_cast<uint8_t>(1u << bitIndex);

    if (active) {
        packedSteps[byteIndex] |= mask;
    } else {
        packedSteps[byteIndex] &= static_cast<uint8_t>(~mask);
    }

    return true;
}

void Pattern::clear() {
    packedSteps[0] = 0;
    packedSteps[1] = 0;
    packedSteps[2] = 0;

    clearRatchets();
}

bool Pattern::setRatchet(uint8_t index, uint8_t code) {
    if (index >= DEFAULT_TOTAL_STEPS || !isValidRatchet(code)) {
        return false;
    }

    const uint8_t byteIndex = index >> 1;      // two nibbles per byte
    const bool highNibble = (index & 0x01) != 0;

    if (highNibble) {
        packedRatchets[byteIndex] =
            static_cast<uint8_t>((packedRatchets[byteIndex] & 0x0F) |
                                 static_cast<uint8_t>(code << 4));
    } else {
        packedRatchets[byteIndex] =
            static_cast<uint8_t>((packedRatchets[byteIndex] & 0xF0) | (code & 0x0F));
    }

    return true;
}

uint8_t Pattern::getRatchet(uint8_t index) const {
    if (index >= DEFAULT_TOTAL_STEPS) {
        return RATCHET_NONE;
    }

    const uint8_t byteIndex = index >> 1;
    const bool highNibble = (index & 0x01) != 0;

    return highNibble ? static_cast<uint8_t>(packedRatchets[byteIndex] >> 4)
                      : static_cast<uint8_t>(packedRatchets[byteIndex] & 0x0F);
}

void Pattern::clearRatchets() {
    for (uint8_t i = 0; i < 12; ++i) {
        packedRatchets[i] = 0;
    }
}

} // namespace flexseq

File: sim/src/domain/Pattern.ts

export const RATCHET_NONE = 0;
export const RATCHET_2 = 2;
export const RATCHET_3 = 3;
export const RATCHET_4 = 4;
export const RATCHET_6 = 6;
export const RATCHET_TRIPLET = 7;

export const RATCHET_CODES: readonly number[] = [
  RATCHET_NONE, RATCHET_2, RATCHET_3, RATCHET_4, RATCHET_6, RATCHET_TRIPLET,
];

export function isValidRatchet(code: number): boolean {
  return RATCHET_CODES.includes(code);
}

export const MIN_SLOT_TICKS = 2;

export function ratchetFitsStep(code: number, ticksPerStep: number): boolean {
  const triggers = ratchetTriggers(code);
  if (triggers <= 1) return true;
  const stepTicks = ticksPerStep * ratchetSpan(code);
  return Math.floor(stepTicks / triggers) >= MIN_SLOT_TICKS;
}

export function ratchetTriggers(code: number): number {
  if (code === RATCHET_TRIPLET) return 3;
  if (code === RATCHET_2 || code === RATCHET_3 || code === RATCHET_4 || code === RATCHET_6) {
    return code;
  }
  return 1;
}

export function ratchetSpan(code: number): number {
  return code === RATCHET_TRIPLET ? 2 : 1;
}

export function ratchetLabel(code: number): string {
  if (code === RATCHET_TRIPLET) return "3T";
  if (code === RATCHET_NONE) return "";
  return String(code);
}
export class Pattern {
  static readonly DEFAULT_TOTAL_STEPS = 36;

  private readonly steps: boolean[];
  private readonly ratchets: number[];

  constructor() {
    this.steps = new Array<boolean>(Pattern.DEFAULT_TOTAL_STEPS).fill(false);
    this.ratchets = new Array<number>(Pattern.DEFAULT_TOTAL_STEPS).fill(RATCHET_NONE);
  }

  private static isValidIndex(index: number): boolean {
    return Number.isInteger(index) && index >= 0 && index < Pattern.DEFAULT_TOTAL_STEPS;
  }

    readStep(index: number): boolean | null {
    if (!Pattern.isValidIndex(index)) return null;
    return this.steps[index] ?? false;
  }

    writeStep(index: number, active: boolean): boolean {
    if (!Pattern.isValidIndex(index)) return false;
    this.steps[index] = active;
    return true;
  }

      setLowStepMask(bits: number): void {
    for (let step = 0; step < 16; ++step) {
      this.steps[step] = ((bits >> step) & 1) === 1;
    }
  }

  clear(): void {
    this.steps.fill(false);
    this.clearRatchets();
  }

      getRatchet(index: number): number {
    if (!Pattern.isValidIndex(index)) return RATCHET_NONE;
    return this.ratchets[index] ?? RATCHET_NONE;
  }

    setRatchet(index: number, code: number): boolean {
    if (!Pattern.isValidIndex(index) || !isValidRatchet(code)) return false;
    this.ratchets[index] = code;
    return true;
  }

  clearRatchets(): void {
    this.ratchets.fill(RATCHET_NONE);
  }

}
